### PR TITLE
[#78] Add AccessBox scheme

### LINKED ISSUE #78: Add authbox scheme for S3 Gateway

### PR DESCRIPTION
**Pull request by KirillovDenis**
close #78 
Signed-off-by: Denis Kirillov <denis@nspcc.ru>

### CODE REVIEW

File: 04-gates/00-gates.md [hunk 1]
```diff
@@ -2,4 +2,54 @@
 # Protocol gateways
 ## HTTP
 ## S3
+
+[NeoFS S3 gateway](https://github.com/nspcc-dev/neofs-s3-gw) provides API compatible with Amazon S3 cloud storage service.
+
+### Access Box scheme
+
+S3 gateway have to authenticate user requests staying in [AWS spec](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html).
```
Comment (EliChin) [inline 04-gates/00-gates.md:10]: S3 gateway has to authenticate user requests regarding

File: 04-gates/00-gates.md [hunk 2]
```diff
@@ -2,4 +2,54 @@
...
+So we have the following scheme:
+
+![Authmate scheme](pic/authmate-scheme.png)
+
+
+1. User use the [neofs-s3-authmate](https://github.com/nspcc-dev/neofs-s3-gw/blob/master/docs/authmate.md)
```
Comment (EliChin) [inline 04-gates/00-gates.md:16]: A user uses

File: 04-gates/00-gates.md [hunk 3]
```diff
@@ -2,4 +2,54 @@
...
+(Authmate) tool to get credentials (`access_key_id` and `secret_access_key`).
+2. Authmate forms "AccessBox" object (see the next section) and put it into NeoFS.
```
Comment (EliChin) [inline 04-gates/00-gates.md:18]: Authmate forms an "AccessBox" object (see the next section) and puts it into NeoFS.

File: 04-gates/00-gates.md [hunk 4]
```diff
@@ -2,4 +2,54 @@
...
+3. Authmate output contains credentials (`access_key_id` and `secret_access_key`) that can be used with AWS CLI for example.
```
Comment (EliChin) [inline 04-gates/00-gates.md:19]: AWS CLI, for example.

File: 04-gates/00-gates.md [hunk 5]
```diff
@@ -2,4 +2,54 @@
...
+4. User send request to NeoFS S3 Gateway using standard AWS tool.
```
Comment (EliChin) [inline 04-gates/00-gates.md:20]: 4. The user sends a request

File: 04-gates/00-gates.md [hunk 6]
```diff
@@ -2,4 +2,54 @@
...
+5. S3 Gateway get "AccessBox" from NeoFS by `access_key_id` and fetch Bearer Token from it.
```
Comment (EliChin) [inline 04-gates/00-gates.md:21]: S3 Gateway gets ... and fetches

File: 04-gates/00-gates.md [hunk 7]
```diff
@@ -2,4 +2,54 @@
...
+6. S3 Gateway use fetched Bearer token to send a request to NeoFS on behalf of user.
```
Comment (EliChin) [inline 04-gates/00-gates.md:22]: uses the fetched ... of the user

File: 04-gates/00-gates.md [hunk 8]
```diff
@@ -2,4 +2,54 @@
...
+
+#### Form Access Box object
+
+Actually, `AccessBox` is a regular object in NeoFS but properly formed. 
+It contains encrypted  Bearer token (`BT`), Session tokens (`STs`) and `secret_access_key`.
```
Comment (EliChin) [inline 04-gates/00-gates.md:27]: an encrypted Bearer Token ... a `secret_access_key`.

File: 04-gates/00-gates.md [hunk 9]
```diff
@@ -2,4 +2,54 @@
...
+
+The credentials are formed by following steps:
```
Comment (EliChin) [inline 04-gates/00-gates.md:29]: by the following

File: 04-gates/00-gates.md [hunk 10]
```diff
@@ -2,4 +2,54 @@
...
+
+1. User provides to Authmate `BT`, `STs` and `s3gw_public_key_k`(public keys of gates that will be able
```
Comment (EliChin) [inline 04-gates/00-gates.md:31]: provides Authmate with `BT` ... (the public keys of the gates that will be able

File: 04-gates/00-gates.md [hunk 11]
```diff
@@ -2,4 +2,54 @@
...
+
+#### Form Access Box object
+
+Actually, `AccessBox` is a regular object in NeoFS but properly formed. 
+It contains encrypted  Bearer token (`BT`), Session tokens (`STs`) and `secret_access_key`.
+
+The credentials are formed by following steps:
+
+1. User provides to Authmate `BT`, `STs` and `s3gw_public_key_k`(public keys of gates that will be able 
+to handle credentials) that will be used when user sends requests via S3 Gateway.
```
Comment (EliChin) [inline 04-gates/00-gates.md:32]: when the user

File: 04-gates/00-gates.md [hunk 12]
```diff
@@ -2,4 +2,54 @@
...
+2. Authmate:
+   1. Generates `secret_access_key` (it's 32 random bytes), and P256 (secp256r1) key pair
```
Comment (EliChin) [inline 04-gates/00-gates.md:34]: Generates a `secret_access_key` (it's 32 random bytes) and a P256 (secp256r1) key pair

File: 04-gates/00-gates.md [hunk 13]
```diff
@@ -2,4 +2,54 @@
...
+It contains encrypted  Bearer token (`BT`), Session tokens (`STs`) and `secret_access_key`.
+
+The credentials are formed by following steps:
+
+1. User provides to Authmate `BT`, `STs` and `s3gw_public_key_k`(public keys of gates that will be able 
+to handle credentials) that will be used when user sends requests via S3 Gateway.
+2. Authmate:
+   1. Generates `secret_access_key` (it's 32 random bytes), and P256 (secp256r1) key pair 
+      (`authmate_private_key`, `authmate_public_key`).
+   2. Form `tokens` protobuf struct that contains `BT`, `STs`, `secret_access_key`
```
Comment (EliChin) [inline 04-gates/00-gates.md:36]: Forms a `tokens` protobuf struct

File: 04-gates/00-gates.md [hunk 14]
```diff
@@ -2,4 +2,54 @@
...
+   3. For each `s3gw_public_key_k` derives symmetric key `symmetric_key_k`
```
Comment (EliChin) [inline 04-gates/00-gates.md:37]: derives a symmetric key

File: 04-gates/00-gates.md [hunk 15]
```diff
@@ -2,4 +2,54 @@
...
+      using [ECDH](https://en.wikipedia.org/wiki/Elliptic-curve_Diffie%E2%80%93Hellman) and 
+      encrypt `tokens` (`encrypted_tokens_k = encrypt(tokens.to_bytes(), secret_key_k, nonce_k)`).
```
Comment (EliChin) [inline 04-gates/00-gates.md:39]: encrypts

File: 04-gates/00-gates.md [hunk 16]
```diff
@@ -2,4 +2,54 @@
...
+   4. For each `encypted_tokens_k` forms struct `gate_k` that contains `encrypted_tokens_k` and `s3gw_public_key_k`.
```
Comment (EliChin) [inline 04-gates/00-gates.md:40]: a struct called `gate_k`

File: 04-gates/00-gates.md [hunk 17]
```diff
@@ -2,4 +2,54 @@
...
+   5. Form final binary object `AccessBox` that contains `authmate_public_key`, `gate_1`, ..., `gate_k`.
```
Comment (EliChin) [inline 04-gates/00-gates.md:41]: Forms the final

File: 04-gates/00-gates.md [hunk 18]
```diff
@@ -2,4 +2,54 @@
...
+   6. Put `AccessBox` object to NeoFS and save its address `CID/OID` as `access_key_id`.
```
Comment (EliChin) [inline 04-gates/00-gates.md:42]: Puts the `AccessBox` object to NeoFS and saves its address

File: 04-gates/00-gates.md [hunk 19]
```diff
@@ -2,4 +2,54 @@
...
+   7. Return pair (`access_key_id`, `access_secret_key`) to user.
```
Comment (EliChin) [inline 04-gates/00-gates.md:43]: Returns the pair (`access_key_id`, `access_secret_key`) to the user.

File: 04-gates/00-gates.md [hunk 20]
```diff
@@ -2,4 +2,54 @@
...
+
+#### Handle S3 request
+On getting request S3 Gateway:
```
Comment (EliChin) [inline 04-gates/00-gates.md:46]: On getting a request, S3 Gateway:

File: 04-gates/00-gates.md [hunk 21]
```diff
@@ -2,4 +2,54 @@
...
+1. Fetch `access_key_id` from `Authorization` header.
```
Comment (EliChin) [inline 04-gates/00-gates.md:47]: Fetches the `access_key_id` from the `Authorization` header.

File: 04-gates/00-gates.md [hunk 22]
```diff
@@ -2,4 +2,54 @@
...
+2. Get `AccessBox` from NeoFS by address (recall `access_key_id` is `CID/OID`)
```
Comment (EliChin) [inline 04-gates/00-gates.md:48]: Gets

File: 04-gates/00-gates.md [hunk 23]
```diff
@@ -2,4 +2,54 @@
...
+3. Using `s3gw_private_key_k` and `authmate_public_key` derives `symmetric_key_k` and decrypts `encrypted_token_k` that 
+was found by matching `s3gw_public_key_k` in `gate_k` struct.
```
Comment (EliChin) [inline 04-gates/00-gates.md:50]: has been found

File: 04-gates/00-gates.md [hunk 24]
```diff
@@ -2,4 +2,54 @@
...
+4. Check signature of initial request using `secret_access_key` from `tokens` struct.
```
Comment (EliChin) [inline 04-gates/00-gates.md:51]: Checks the signature of the initial request

File: 04-gates/00-gates.md [hunk 25]
```diff
@@ -2,4 +2,54 @@
...
+5. Use `BT` and `STs` to perform requests to NeoFS.
```
Comment (EliChin) [inline 04-gates/00-gates.md:52]: Uses

**Review by EliChin — CHANGES_REQUESTED** [review]
Please, pay attention to the English fixes suggested

**Review by EliChin — APPROVED** [review]

**Review by realloc — APPROVED** [review]

**Pull request merged**